// vidstack/dist/prod/chunks/vidstack-271b581f.js
import { H as Host, e as effect, f as setAttribute, j as computed } from './vidstack-fd2ab5eb.js';
import { u as useMediaContext, M as MediaProvider, a as MediaPlayer } from './vidstack-01cf3c93.js';

class MediaProviderElement extends Host(HTMLElement, MediaProvider) {
  constructor() {
    super(...arguments);
    this.D = null;
  }
  static {
    this.tagName = "media-provider";
  }
  onSetup() {
    this.g = useMediaContext();
    this.setAttribute("keep-alive", "");
  }
  onDestroy() {
    this.D?.remove();
    this.D = null;
  }
  onConnect() {
    effect(() => {
      const loader = this.$state.loader();
      const media = loader ? loader.mediaType() === "audio" ? this.E() : this.F() : null;
      if (this.D !== media) {
        const parent = this.D?.parentElement ?? this;
        this.D?.remove();
        this.D = media;
        if (media)
          parent.prepend(media);
      }
      this.load(media);
    });
  }
  E() {
    const audio = this.D instanceof HTMLAudioElement ? this.D : document.createElement("audio");
    setAttribute(audio, "preload", "none");
    setAttribute(audio, "aria-hidden", "true");
    const { controls, crossorigin } = this.g.$state;
    effect(() => {
      setAttribute(audio, "controls", controls());
      setAttribute(audio, "crossorigin", crossorigin());
    });
    return audio;
  }
  F() {
    const video = this.D instanceof HTMLVideoElement ? this.D : document.createElement("video");
    const { controls, crossorigin, poster } = this.g.$state, $poster = computed(() => poster() && controls() ? poster() : null);
    effect(() => {
      setAttribute(video, "controls", controls());
      setAttribute(video, "crossorigin", crossorigin());
      setAttribute(video, "poster", $poster());
    });
    return video;
  }
}

class MediaPlayerElement extends Host(HTMLElement, MediaPlayer) {
  static {
    this.tagName = "media-player";
  }
  static {
    this.attrs = {
      preferNativeHLS: "prefer-native-hls"
    };
  }
}

export { MediaProviderElement as M, MediaPlayerElement as a };
