// vidstack/dist/prod/chunks/vidstack-26ec82d8.js
/**
 * @license
 * Copyright 2017 Google LLC
 * SPDX-License-Identifier: BSD-3-Clause
 */
var t$2;
const i$1 = window, s$1 = i$1.trustedTypes, e$3 = s$1 ? s$1.createPolicy("lit-html", { createHTML: (t2) => t2 }) : void 0, o$2 = "$lit$", n$1 = `lit$${(Math.random() + "").slice(9)}$`, l$2 = "?" + n$1, h$1 = `<${l$2}>`, r$1 = document, u = () => r$1.createComment(""), d = (t2) => null === t2 || "object" != typeof t2 && "function" != typeof t2, c$1 = Array.isArray, v = (t2) => c$1(t2) || "function" == typeof (null == t2 ? void 0 : t2[Symbol.iterator]), a = "[ 	\n\f\r]", f = /<(?:(!--|\/[^a-zA-Z])|(\/?[a-zA-Z][^>\s]*)|(\/?$))/g, _ = /-->/g, m = />/g, p = RegExp(`>|${a}(?:([^\\s"'>=/]+)(${a}*=${a}*(?:[^ 	
\f\r"'\`<>=]|("|')|))|$)`, "g"), g = /'/g, $ = /"/g, y = /^(?:script|style|textarea|title)$/i, w = (t2) => (i2, ...s2) => ({ gh: t2, strings: i2, values: s2 }), x = w(1), T = Symbol.for("lit-noChange"), A = Symbol.for("lit-nothing"), E = /* @__PURE__ */ new WeakMap(), C = r$1.createTreeWalker(r$1, 129, null, false);
function P(t2, i2) {
  if (!Array.isArray(t2) || !t2.hasOwnProperty("raw"))
    throw Error("invalid template strings array");
  return void 0 !== e$3 ? e$3.createHTML(i2) : i2;
}
const V = (t2, i2) => {
  const s2 = t2.length - 1, e2 = [];
  let l2, r2 = 2 === i2 ? "<svg>" : "", u2 = f;
  for (let i3 = 0; i3 < s2; i3++) {
    const s3 = t2[i3];
    let d2, c2, v2 = -1, a2 = 0;
    for (; a2 < s3.length && (u2.lastIndex = a2, c2 = u2.exec(s3), null !== c2); )
      a2 = u2.lastIndex, u2 === f ? "!--" === c2[1] ? u2 = _ : void 0 !== c2[1] ? u2 = m : void 0 !== c2[2] ? (y.test(c2[2]) && (l2 = RegExp("</" + c2[2], "g")), u2 = p) : void 0 !== c2[3] && (u2 = p) : u2 === p ? ">" === c2[0] ? (u2 = null != l2 ? l2 : f, v2 = -1) : void 0 === c2[1] ? v2 = -2 : (v2 = u2.lastIndex - c2[2].length, d2 = c2[1], u2 = void 0 === c2[3] ? p : '"' === c2[3] ? $ : g) : u2 === $ || u2 === g ? u2 = p : u2 === _ || u2 === m ? u2 = f : (u2 = p, l2 = void 0);
    const w2 = u2 === p && t2[i3 + 1].startsWith("/>") ? " " : "";
    r2 += u2 === f ? s3 + h$1 : v2 >= 0 ? (e2.push(d2), s3.slice(0, v2) + o$2 + s3.slice(v2) + n$1 + w2) : s3 + n$1 + (-2 === v2 ? (e2.push(void 0), i3) : w2);
  }
  return [P(t2, r2 + (t2[s2] || "<?>") + (2 === i2 ? "</svg>" : "")), e2];
};
class N {
  constructor({ strings: t2, gh: i2 }, e2) {
    let h2;
    this.parts = [];
    let r2 = 0, d2 = 0;
    const c2 = t2.length - 1, v2 = this.parts, [a2, f2] = V(t2, i2);
    if (this.el = N.createElement(a2, e2), C.currentNode = this.el.content, 2 === i2) {
      const t3 = this.el.content, i3 = t3.firstChild;
      i3.remove(), t3.append(...i3.childNodes);
    }
    for (; null !== (h2 = C.nextNode()) && v2.length < c2; ) {
      if (1 === h2.nodeType) {
        if (h2.hasAttributes()) {
          const t3 = [];
          for (const i3 of h2.getAttributeNames())
            if (i3.endsWith(o$2) || i3.startsWith(n$1)) {
              const s2 = f2[d2++];
              if (t3.push(i3), void 0 !== s2) {
                const t4 = h2.getAttribute(s2.toLowerCase() + o$2).split(n$1), i4 = /([.?@])?(.*)/.exec(s2);
                v2.push({ type: 1, index: r2, name: i4[2], strings: t4, ctor: "." === i4[1] ? H : "?" === i4[1] ? L : "@" === i4[1] ? z : k });
              } else
                v2.push({ type: 6, index: r2 });
            }
          for (const i3 of t3)
            h2.removeAttribute(i3);
        }
        if (y.test(h2.tagName)) {
          const t3 = h2.textContent.split(n$1), i3 = t3.length - 1;
          if (i3 > 0) {
            h2.textContent = s$1 ? s$1.emptyScript : "";
            for (let s2 = 0; s2 < i3; s2++)
              h2.append(t3[s2], u()), C.nextNode(), v2.push({ type: 2, index: ++r2 });
            h2.append(t3[i3], u());
          }
        }
      } else if (8 === h2.nodeType)
        if (h2.data === l$2)
          v2.push({ type: 2, index: r2 });
        else {
          let t3 = -1;
          for (; -1 !== (t3 = h2.data.indexOf(n$1, t3 + 1)); )
            v2.push({ type: 7, index: r2 }), t3 += n$1.length - 1;
        }
      r2++;
    }
  }
  static createElement(t2, i2) {
    const s2 = r$1.createElement("template");
    return s2.innerHTML = t2, s2;
  }
}
function S(t2, i2, s2 = t2, e2) {
  var o2, n2, l2, h2;
  if (i2 === T)
    return i2;
  let r2 = void 0 !== e2 ? null === (o2 = s2.ih) || void 0 === o2 ? void 0 : o2[e2] : s2.nh;
  const u2 = d(i2) ? void 0 : i2.rh;
  return (null == r2 ? void 0 : r2.constructor) !== u2 && (null === (n2 = null == r2 ? void 0 : r2.sh) || void 0 === n2 || n2.call(r2, false), void 0 === u2 ? r2 = void 0 : (r2 = new u2(t2), r2.th(t2, s2, e2)), void 0 !== e2 ? (null !== (l2 = (h2 = s2).ih) && void 0 !== l2 ? l2 : h2.ih = [])[e2] = r2 : s2.nh = r2), void 0 !== r2 && (i2 = S(t2, r2.uh(t2, i2.values), r2, e2)), i2;
}
class M {
  constructor(t2, i2) {
    this.jh = [], this.hh = void 0, this.kh = t2, this.ah = i2;
  }
  get parentNode() {
    return this.ah.parentNode;
  }
  get bh() {
    return this.ah.bh;
  }
  u(t2) {
    var i2;
    const { el: { content: s2 }, parts: e2 } = this.kh, o2 = (null !== (i2 = null == t2 ? void 0 : t2.creationScope) && void 0 !== i2 ? i2 : r$1).importNode(s2, true);
    C.currentNode = o2;
    let n2 = C.nextNode(), l2 = 0, h2 = 0, u2 = e2[0];
    for (; void 0 !== u2; ) {
      if (l2 === u2.index) {
        let i3;
        2 === u2.type ? i3 = new R(n2, n2.nextSibling, this, t2) : 1 === u2.type ? i3 = new u2.ctor(n2, u2.name, u2.strings, this, t2) : 6 === u2.type && (i3 = new Z(n2, this, t2)), this.jh.push(i3), u2 = e2[++h2];
      }
      l2 !== (null == u2 ? void 0 : u2.index) && (n2 = C.nextNode(), l2++);
    }
    return C.currentNode = r$1, o2;
  }
  v(t2) {
    let i2 = 0;
    for (const s2 of this.jh)
      void 0 !== s2 && (void 0 !== s2.strings ? (s2.ch(t2, s2, i2), i2 += s2.strings.length - 2) : s2.ch(t2[i2])), i2++;
  }
}
class R {
  constructor(t2, i2, s2, e2) {
    var o2;
    this.type = 2, this.$g = A, this.hh = void 0, this.dh = t2, this.eh = i2, this.ah = s2, this.options = e2, this.lh = null === (o2 = null == e2 ? void 0 : e2.isConnected) || void 0 === o2 || o2;
  }
  get bh() {
    var t2, i2;
    return null !== (i2 = null === (t2 = this.ah) || void 0 === t2 ? void 0 : t2.bh) && void 0 !== i2 ? i2 : this.lh;
  }
  get parentNode() {
    let t2 = this.dh.parentNode;
    const i2 = this.ah;
    return void 0 !== i2 && 11 === (null == t2 ? void 0 : t2.nodeType) && (t2 = i2.parentNode), t2;
  }
  get startNode() {
    return this.dh;
  }
  get endNode() {
    return this.eh;
  }
  ch(t2, i2 = this) {
    t2 = S(this, t2, i2), d(t2) ? t2 === A || null == t2 || "" === t2 ? (this.$g !== A && this.fh(), this.$g = A) : t2 !== this.$g && t2 !== T && this.mh(t2) : void 0 !== t2.gh ? this.g(t2) : void 0 !== t2.nodeType ? this.$(t2) : v(t2) ? this.T(t2) : this.mh(t2);
  }
  k(t2) {
    return this.dh.parentNode.insertBefore(t2, this.eh);
  }
  $(t2) {
    this.$g !== t2 && (this.fh(), this.$g = this.k(t2));
  }
  mh(t2) {
    this.$g !== A && d(this.$g) ? this.dh.nextSibling.data = t2 : this.$(r$1.createTextNode(t2)), this.$g = t2;
  }
  g(t2) {
    var i2;
    const { values: s2, gh: e2 } = t2, o2 = "number" == typeof e2 ? this.oh(t2) : (void 0 === e2.el && (e2.el = N.createElement(P(e2.h, e2.h[0]), this.options)), e2);
    if ((null === (i2 = this.$g) || void 0 === i2 ? void 0 : i2.kh) === o2)
      this.$g.v(s2);
    else {
      const t3 = new M(o2, this), i3 = t3.u(this.options);
      t3.v(s2), this.$(i3), this.$g = t3;
    }
  }
  oh(t2) {
    let i2 = E.get(t2.strings);
    return void 0 === i2 && E.set(t2.strings, i2 = new N(t2)), i2;
  }
  T(t2) {
    c$1(this.$g) || (this.$g = [], this.fh());
    const i2 = this.$g;
    let s2, e2 = 0;
    for (const o2 of t2)
      e2 === i2.length ? i2.push(s2 = new R(this.k(u()), this.k(u()), this, this.options)) : s2 = i2[e2], s2.ch(o2), e2++;
    e2 < i2.length && (this.fh(s2 && s2.eh.nextSibling, e2), i2.length = e2);
  }
  fh(t2 = this.dh.nextSibling, i2) {
    var s2;
    for (null === (s2 = this.ph) || void 0 === s2 || s2.call(this, false, true, i2); t2 && t2 !== this.eh; ) {
      const i3 = t2.nextSibling;
      t2.remove(), t2 = i3;
    }
  }
  setConnected(t2) {
    var i2;
    void 0 === this.ah && (this.lh = t2, null === (i2 = this.ph) || void 0 === i2 || i2.call(this, t2));
  }
}
class k {
  constructor(t2, i2, s2, e2, o2) {
    this.type = 1, this.$g = A, this.hh = void 0, this.element = t2, this.name = i2, this.ah = e2, this.options = o2, s2.length > 2 || "" !== s2[0] || "" !== s2[1] ? (this.$g = Array(s2.length - 1).fill(new String()), this.strings = s2) : this.$g = A;
  }
  get tagName() {
    return this.element.tagName;
  }
  get bh() {
    return this.ah.bh;
  }
  ch(t2, i2 = this, s2, e2) {
    const o2 = this.strings;
    let n2 = false;
    if (void 0 === o2)
      t2 = S(this, t2, i2, 0), n2 = !d(t2) || t2 !== this.$g && t2 !== T, n2 && (this.$g = t2);
    else {
      const e3 = t2;
      let l2, h2;
      for (t2 = o2[0], l2 = 0; l2 < o2.length - 1; l2++)
        h2 = S(this, e3[s2 + l2], i2, l2), h2 === T && (h2 = this.$g[l2]), n2 || (n2 = !d(h2) || h2 !== this.$g[l2]), h2 === A ? t2 = A : t2 !== A && (t2 += (null != h2 ? h2 : "") + o2[l2 + 1]), this.$g[l2] = h2;
    }
    n2 && !e2 && this.j(t2);
  }
  j(t2) {
    t2 === A ? this.element.removeAttribute(this.name) : this.element.setAttribute(this.name, null != t2 ? t2 : "");
  }
}
class H extends k {
  constructor() {
    super(...arguments), this.type = 3;
  }
  j(t2) {
    this.element[this.name] = t2 === A ? void 0 : t2;
  }
}
const I = s$1 ? s$1.emptyScript : "";
class L extends k {
  constructor() {
    super(...arguments), this.type = 4;
  }
  j(t2) {
    t2 && t2 !== A ? this.element.setAttribute(this.name, I) : this.element.removeAttribute(this.name);
  }
}
class z extends k {
  constructor(t2, i2, s2, e2, o2) {
    super(t2, i2, s2, e2, o2), this.type = 5;
  }
  ch(t2, i2 = this) {
    var s2;
    if ((t2 = null !== (s2 = S(this, t2, i2, 0)) && void 0 !== s2 ? s2 : A) === T)
      return;
    const e2 = this.$g, o2 = t2 === A && e2 !== A || t2.capture !== e2.capture || t2.once !== e2.once || t2.passive !== e2.passive, n2 = t2 !== A && (e2 === A || o2);
    o2 && this.element.removeEventListener(this.name, this, e2), n2 && this.element.addEventListener(this.name, this, t2), this.$g = t2;
  }
  handleEvent(t2) {
    var i2, s2;
    "function" == typeof this.$g ? this.$g.call(null !== (s2 = null === (i2 = this.options) || void 0 === i2 ? void 0 : i2.host) && void 0 !== s2 ? s2 : this.element, t2) : this.$g.handleEvent(t2);
  }
}
class Z {
  constructor(t2, i2, s2) {
    this.element = t2, this.type = 6, this.hh = void 0, this.ah = i2, this.options = s2;
  }
  get bh() {
    return this.ah.bh;
  }
  ch(t2) {
    S(this, t2);
  }
}
const B = i$1.litHtmlPolyfillSupport;
null == B || B(N, R), (null !== (t$2 = i$1.litHtmlVersions) && void 0 !== t$2 ? t$2 : i$1.litHtmlVersions = []).push("2.8.0");
const D = (t2, i2, s2) => {
  var e2, o2;
  const n2 = null !== (e2 = null == s2 ? void 0 : s2.renderBefore) && void 0 !== e2 ? e2 : i2;
  let l2 = n2.qh;
  if (void 0 === l2) {
    const t3 = null !== (o2 = null == s2 ? void 0 : s2.renderBefore) && void 0 !== o2 ? o2 : null;
    n2.qh = l2 = new R(i2.insertBefore(u(), t3), t3, void 0, null != s2 ? s2 : {});
  }
  return l2.ch(t2), l2;
};

/**
 * @license
 * Copyright 2020 Google LLC
 * SPDX-License-Identifier: BSD-3-Clause
 */
const e$2 = (o2) => void 0 === o2.strings;

/**
 * @license
 * Copyright 2017 Google LLC
 * SPDX-License-Identifier: BSD-3-Clause
 */
const t$1 = { ATTRIBUTE: 1, CHILD: 2, PROPERTY: 3, BOOLEAN_ATTRIBUTE: 4, EVENT: 5, ELEMENT: 6 }, e$1 = (t2) => (...e2) => ({ rh: t2, values: e2 });
class i {
  constructor(t2) {
  }
  get bh() {
    return this.ah.bh;
  }
  th(t2, e2, i2) {
    this.Gj = t2, this.ah = e2, this.Ij = i2;
  }
  uh(t2, e2) {
    return this.update(t2, e2);
  }
  update(t2, e2) {
    return this.render(...e2);
  }
}

/**
 * @license
 * Copyright 2017 Google LLC
 * SPDX-License-Identifier: BSD-3-Clause
 */
const s = (i2, t2) => {
  var e2, o2;
  const r2 = i2.hh;
  if (void 0 === r2)
    return false;
  for (const i3 of r2)
    null === (o2 = (e2 = i3).sh) || void 0 === o2 || o2.call(e2, t2, false), s(i3, t2);
  return true;
}, o$1 = (i2) => {
  let t2, e2;
  do {
    if (void 0 === (t2 = i2.ah))
      break;
    e2 = t2.hh, e2.delete(i2), i2 = t2;
  } while (0 === (null == e2 ? void 0 : e2.size));
}, r = (i2) => {
  for (let t2; t2 = i2.ah; i2 = t2) {
    let e2 = t2.hh;
    if (void 0 === e2)
      t2.hh = e2 = /* @__PURE__ */ new Set();
    else if (e2.has(i2))
      break;
    e2.add(i2), l$1(t2);
  }
};
function n(i2) {
  void 0 !== this.hh ? (o$1(this), this.ah = i2, r(this)) : this.ah = i2;
}
function h(i2, t2 = false, e2 = 0) {
  const r2 = this.$g, n2 = this.hh;
  if (void 0 !== n2 && 0 !== n2.size)
    if (t2)
      if (Array.isArray(r2))
        for (let i3 = e2; i3 < r2.length; i3++)
          s(r2[i3], false), o$1(r2[i3]);
      else
        null != r2 && (s(r2, false), o$1(r2));
    else
      s(this, i2);
}
const l$1 = (i2) => {
  var t2, s2, o2, r2;
  i2.type == t$1.CHILD && (null !== (t2 = (o2 = i2).ph) && void 0 !== t2 || (o2.ph = h), null !== (s2 = (r2 = i2).Hj) && void 0 !== s2 || (r2.Hj = n));
};
class c extends i {
  constructor() {
    super(...arguments), this.hh = void 0;
  }
  th(i2, t2, e2) {
    super.th(i2, t2, e2), r(this), this.isConnected = i2.bh;
  }
  sh(i2, t2 = true) {
    var e2, r2;
    i2 !== this.isConnected && (this.isConnected = i2, i2 ? null === (e2 = this.reconnected) || void 0 === e2 || e2.call(this) : null === (r2 = this.disconnected) || void 0 === r2 || r2.call(this)), t2 && (s(this, i2), o$1(this));
  }
  setValue(t2) {
    if (e$2(this.Gj))
      this.Gj.ch(t2, this);
    else {
      const i2 = [...this.Gj.$g];
      i2[this.Ij] = t2, this.Gj.ch(i2, this, 0);
    }
  }
  disconnected() {
  }
  reconnected() {
  }
}

/**
 * @license
 * Copyright 2018 Google LLC
 * SPDX-License-Identifier: BSD-3-Clause
 */
const l = (l2) => null != l2 ? l2 : A;

/**
 * @license
 * Copyright 2017 Google LLC
 * SPDX-License-Identifier: BSD-3-Clause
 */
class e extends i {
  constructor(i2) {
    if (super(i2), this.et = A, i2.type !== t$1.CHILD)
      throw Error(this.constructor.directiveName + "() can only be used in child bindings");
  }
  render(r2) {
    if (r2 === A || null == r2)
      return this.ft = void 0, this.et = r2;
    if (r2 === T)
      return r2;
    if ("string" != typeof r2)
      throw Error(this.constructor.directiveName + "() called with a non-string value");
    if (r2 === this.et)
      return this.ft;
    this.et = r2;
    const s2 = [r2];
    return s2.raw = s2, this.ft = { gh: this.constructor.resultType, strings: s2, values: [] };
  }
}
e.directiveName = "unsafeHTML", e.resultType = 1;

/**
 * @license
 * Copyright 2017 Google LLC
 * SPDX-License-Identifier: BSD-3-Clause
 */
class t extends e {
}
t.directiveName = "unsafeSVG", t.resultType = 2;
const o = e$1(t);

export { D, c, e$1 as e, l, o, t$1 as t, x };
