// vidstack/dist/dev/chunks/vidstack-af302770.js
import { H as Host, e as effect, f as setAttribute, j as computed } from './vidstack-b8ba7e3c.js';
import { u as useMediaContext, M as MediaProvider, a as MediaPlayer } from './vidstack-1c6c4f6e.js';

class MediaProviderElement extends Host(HTMLElement, MediaProvider) {
  constructor() {
    super(...arguments);
    this._mediaElement = null;
  }
  static {
    this.tagName = "media-provider";
  }
  onSetup() {
    this._media = useMediaContext();
    this.setAttribute("keep-alive", "");
  }
  onDestroy() {
    this._mediaElement?.remove();
    this._mediaElement = null;
  }
  onConnect() {
    effect(() => {
      const loader = this.$state.loader();
      const media = loader ? loader.mediaType() === "audio" ? this._createAudio() : this._createVideo() : null;
      if (this._mediaElement !== media) {
        const parent = this._mediaElement?.parentElement ?? this;
        this._mediaElement?.remove();
        this._mediaElement = media;
        if (media)
          parent.prepend(media);
      }
      this.load(media);
    });
  }
  _createAudio() {
    const audio = this._mediaElement instanceof HTMLAudioElement ? this._mediaElement : document.createElement("audio");
    setAttribute(audio, "preload", "none");
    setAttribute(audio, "aria-hidden", "true");
    const { controls, crossorigin } = this._media.$state;
    effect(() => {
      setAttribute(audio, "controls", controls());
      setAttribute(audio, "crossorigin", crossorigin());
    });
    return audio;
  }
  _createVideo() {
    const video = this._mediaElement instanceof HTMLVideoElement ? this._mediaElement : document.createElement("video");
    const { controls, crossorigin, poster } = this._media.$state, $poster = computed(() => poster() && controls() ? poster() : null);
    effect(() => {
      setAttribute(video, "controls", controls());
      setAttribute(video, "crossorigin", crossorigin());
      setAttribute(video, "poster", $poster());
    });
    return video;
  }
}

class MediaPlayerElement extends Host(HTMLElement, MediaPlayer) {
  static {
    this.tagName = "media-player";
  }
  static {
    this.attrs = {
      preferNativeHLS: "prefer-native-hls"
    };
  }
}

export { MediaProviderElement as M, MediaPlayerElement as a };
