// vidstack/dist/dev/chunks/vidstack-3484a78d.js
import { H as Host, e as effect, f as setAttribute, h as signal, j as computed, C as Component, l as listenEvent, k as isString, u as useState, N as BOOLEAN } from './vidstack-b8ba7e3c.js';
import { am as Captions, al as Gesture, u as useMediaContext, an as Poster, T as Thumbnail, ao as Time, N as Controls, P as PlayerQueryList, U as PlayButton, Y as MuteButton, W as CaptionButton, X as FullscreenButton, Z as PIPButton, _ as SeekButton, R as ToggleButton, $ as LiveButton, O as Tooltip, Q as TooltipContent, ah as AudioRadioGroup, ai as CaptionsRadioGroup, a9 as Menu, aa as MenuButton, ab as MenuPortal, ad as MenuItems, ag as ChaptersRadioGroup, aj as SpeedRadioGroup, ak as QualityRadioGroup, af as Radio, ae as RadioGroup, S as Slider, a3 as SliderValue, a2 as SliderVideo, a7 as TimeSlider, a4 as SliderPreview, a6 as VolumeSlider, a8 as SliderChapters } from './vidstack-1c6c4f6e.js';
import { C as ControlsGroup, T as TooltipTrigger, M as MenuItem } from './vidstack-5ca95e20.js';
import { o as observeActiveTextTrack } from './vidstack-dd29a7a5.js';
import { c as cloneTemplateContent, a as createTemplate, r as requestScopedAnimationFrame, b as cloneTemplate } from './vidstack-b97c0002.js';

class MediaCaptionsElement extends Host(HTMLElement, Captions) {
  static {
    this.tagName = "media-captions";
  }
}

class MediaGestureElement extends Host(HTMLElement, Gesture) {
  static {
    this.tagName = "media-gesture";
  }
}

class MediaPosterElement extends Host(HTMLElement, Poster) {
  constructor() {
    super(...arguments);
    this._img = document.createElement("img");
  }
  static {
    this.tagName = "media-poster";
  }
  onSetup() {
    this._media = useMediaContext();
    this.$state.img.set(this._img);
  }
  onConnect() {
    const { src, alt } = this.$state, { crossorigin } = this._media.$state;
    if (this._img.parentNode !== this) {
      this.prepend(this._img);
    }
    effect(() => {
      setAttribute(this._img, "src", src());
      setAttribute(this._img, "alt", alt());
      setAttribute(this._img, "crossorigin", crossorigin());
    });
  }
}

const imgTemplate = /* @__PURE__ */ createTemplate(
  '<img loading="eager" decoding="async" aria-hidden="true">'
);
class MediaThumbnailElement extends Host(HTMLElement, Thumbnail) {
  constructor() {
    super(...arguments);
    this._img = this._createImg();
  }
  static {
    this.tagName = "media-thumbnail";
  }
  onSetup() {
    this._media = useMediaContext();
    this.$state.img.set(this._img);
  }
  onConnect() {
    const { src } = this.$state, { crossorigin } = this._media.$props;
    if (this._img.parentNode !== this) {
      this.prepend(this._img);
    }
    effect(() => {
      setAttribute(this._img, "src", src());
      setAttribute(this._img, "crossorigin", crossorigin());
    });
  }
  _createImg() {
    return cloneTemplateContent(imgTemplate);
  }
}

class MediaTimeElement extends Host(HTMLElement, Time) {
  static {
    this.tagName = "media-time";
  }
  onConnect() {
    effect(() => {
      this.textContent = this.$state.timeText();
    });
  }
}

class MediaControlsElement extends Host(HTMLElement, Controls) {
  static {
    this.tagName = "media-controls";
  }
}

class MediaControlsGroupElement extends Host(HTMLElement, ControlsGroup) {
  static {
    this.tagName = "media-controls-group";
  }
}

class ChapterTitle extends Component {
}
class MediaChapterTitleElement extends Host(HTMLElement, ChapterTitle) {
  constructor() {
    super(...arguments);
    this._chapterTitle = signal("");
    this._title = computed(() => {
      const { title, started } = this._media.$state;
      return started() ? this._chapterTitle() || title() : title();
    });
  }
  static {
    this.tagName = "media-chapter-title";
  }
  onSetup() {
    this._media = useMediaContext();
  }
  onConnect() {
    observeActiveTextTrack(this._media.textTracks, "chapters", (track) => {
      if (!track) {
        this._chapterTitle.set("");
        return;
      }
      const onCueChange = () => {
        const activeCue = track?.activeCues[0];
        this._chapterTitle.set(activeCue?.text || "");
      };
      onCueChange();
      listenEvent(track, "cue-change", onCueChange);
    });
    effect(() => {
      this.textContent = this._title();
    });
  }
}

class MediaLayout extends Component {
  static {
    this.props = {
      when: ""
    };
  }
}
class MediaLayoutElement extends Host(HTMLElement, MediaLayout) {
  static {
    this.tagName = "media-layout";
  }
  onSetup() {
    this.query = PlayerQueryList.create(this.$props.when);
  }
  onConnect() {
    effect(this._watchQuery.bind(this));
  }
  _watchQuery() {
    const root = this.firstElementChild, isTemplate = root?.localName === "template", isHTMLElement = root instanceof HTMLElement;
    if (!this.query.matches) {
      if (isTemplate) {
        for (const el of this.children) {
          if (el.localName !== "template")
            el.remove();
        }
      } else if (isHTMLElement) {
        root.style.display = "none";
      }
      return;
    }
    if (isTemplate) {
      this.append(root.cloneNode(true));
    } else if (isHTMLElement) {
      root.style.display = "";
    }
  }
}

class MediaPlayButtonElement extends Host(HTMLElement, PlayButton) {
  static {
    this.tagName = "media-play-button";
  }
}

class MediaMuteButtonElement extends Host(HTMLElement, MuteButton) {
  static {
    this.tagName = "media-mute-button";
  }
}

class MediaCaptionButtonElement extends Host(HTMLElement, CaptionButton) {
  static {
    this.tagName = "media-caption-button";
  }
}

class MediaFullscreenButtonElement extends Host(HTMLElement, FullscreenButton) {
  static {
    this.tagName = "media-fullscreen-button";
  }
}

class MediaPIPButtonElement extends Host(HTMLElement, PIPButton) {
  static {
    this.tagName = "media-pip-button";
  }
}

class MediaSeekButtonElement extends Host(HTMLElement, SeekButton) {
  static {
    this.tagName = "media-seek-button";
  }
}

class MediaToggleButtonElement extends Host(HTMLElement, ToggleButton) {
  static {
    this.tagName = "media-toggle-button";
  }
}

class MediaLiveButtonElement extends Host(HTMLElement, LiveButton) {
  static {
    this.tagName = "media-live-button";
  }
}

class MediaTooltipElement extends Host(HTMLElement, Tooltip) {
  static {
    this.tagName = "media-tooltip";
  }
}

class MediaTooltipTriggerElement extends Host(HTMLElement, TooltipTrigger) {
  static {
    this.tagName = "media-tooltip-trigger";
  }
  onConnect() {
    this.style.display = "contents";
  }
}

class MediaTooltipContentElement extends Host(HTMLElement, TooltipContent) {
  static {
    this.tagName = "media-tooltip-content";
  }
}

function renderMenuItemsTemplate(el, onCreate) {
  requestScopedAnimationFrame(() => {
    if (!el.connectScope)
      return;
    const template = el.querySelector("template");
    if (!template)
      return;
    effect(() => {
      if (!template.content.firstElementChild?.localName && !template.firstElementChild) {
        throw Error("[vidstack] menu items template requires root element");
      }
      const options = el.getOptions();
      cloneTemplate(template, options.length, (radio, i) => {
        const { label, value } = options[i], labelEl = radio.querySelector(`[data-part="label"]`);
        radio.setAttribute("value", value);
        if (labelEl) {
          if (isString(label)) {
            labelEl.textContent = label;
          } else {
            effect(() => {
              labelEl.textContent = label();
            });
          }
        }
        onCreate?.(radio, options[i], i);
      });
    });
  });
}

class MediaAudioRadioGroupElement extends Host(HTMLElement, AudioRadioGroup) {
  static {
    this.tagName = "media-audio-radio-group";
  }
  onConnect() {
    renderMenuItemsTemplate(this);
  }
}

class MediaCaptionsRadioGroupElement extends Host(HTMLElement, CaptionsRadioGroup) {
  static {
    this.tagName = "media-captions-radio-group";
  }
  onConnect() {
    renderMenuItemsTemplate(this);
  }
}

class MediaMenuElement extends Host(HTMLElement, Menu) {
  static {
    this.tagName = "media-menu";
  }
}

class MediaMenuButtonElement extends Host(HTMLElement, MenuButton) {
  static {
    this.tagName = "media-menu-button";
  }
}

class MediaMenuPortalElement extends Host(HTMLElement, MenuPortal) {
  static {
    this.tagName = "media-menu-portal";
  }
  static {
    this.attrs = {
      disabled: {
        converter(value) {
          if (isString(value))
            return value;
          return value !== null;
        }
      }
    };
  }
}

class MediaMenuItemElement extends Host(HTMLElement, MenuItem) {
  static {
    this.tagName = "media-menu-item";
  }
}

class MediaMenuItemsElement extends Host(HTMLElement, MenuItems) {
  static {
    this.tagName = "media-menu-items";
  }
}

class MediaChaptersRadioGroupElement extends Host(HTMLElement, ChaptersRadioGroup) {
  static {
    this.tagName = "media-chapters-radio-group";
  }
  onConnect() {
    renderMenuItemsTemplate(this, (el, option) => {
      const { cue, startTime, duration } = option, thumbnailEl = el.querySelector(".vds-thumbnail,media-thumbnail"), startEl = el.querySelector('[data-part="start-time"]'), durationEl = el.querySelector('[data-part="duration"]');
      if (startEl)
        startEl.textContent = startTime;
      if (durationEl)
        durationEl.textContent = duration;
      if (thumbnailEl) {
        thumbnailEl.setAttribute("time", cue.startTime + "");
        effect(() => {
          const { thumbnails } = this.$props;
          thumbnailEl.setAttribute("src", thumbnails());
        });
      }
    });
  }
}

class MediaSpeedRadioGroupElement extends Host(HTMLElement, SpeedRadioGroup) {
  static {
    this.tagName = "media-speed-radio-group";
  }
  onConnect() {
    renderMenuItemsTemplate(this);
  }
}

class MediaQualityRadioGroupElement extends Host(HTMLElement, QualityRadioGroup) {
  static {
    this.tagName = "media-quality-radio-group";
  }
  onConnect() {
    renderMenuItemsTemplate(this, (el, option) => {
      const bitrate = option.bitrate, bitrateEl = el.querySelector('[data-part="bitrate"]');
      if (bitrate && bitrateEl) {
        effect(() => {
          bitrateEl.textContent = bitrate() || "";
        });
      }
    });
  }
}

class MediaRadioElement extends Host(HTMLElement, Radio) {
  static {
    this.tagName = "media-radio";
  }
}

class MediaRadioGroupElement extends Host(HTMLElement, RadioGroup) {
  static {
    this.tagName = "media-radio-group";
  }
}

class MediaSliderElement extends Host(HTMLElement, Slider) {
  static {
    this.tagName = "media-slider";
  }
}

class MediaSliderThumbnailElement extends MediaThumbnailElement {
  static {
    this.tagName = "media-slider-thumbnail";
  }
  onSetup() {
    super.onSetup();
    this._slider = useState(Slider.state);
  }
  onConnect() {
    super.onConnect();
    effect(this._watchTime.bind(this));
  }
  _watchTime() {
    const { duration } = this._media.$state;
    this.time = this._slider.pointerRate() * duration();
  }
}

class MediaSliderValueElement extends Host(HTMLElement, SliderValue) {
  static {
    this.tagName = "media-slider-value";
  }
  static {
    this.attrs = {
      padMinutes: {
        converter: BOOLEAN
      }
    };
  }
  onConnect() {
    effect(() => {
      this.textContent = this.getValueText();
    });
  }
}

const videoTemplate = /* @__PURE__ */ createTemplate(
  `<video muted playsinline preload="none" style="max-width: unset;"></video>`
);
class MediaSliderVideoElement extends Host(HTMLElement, SliderVideo) {
  constructor() {
    super(...arguments);
    this._video = this._createVideo();
  }
  static {
    this.tagName = "media-slider-video";
  }
  onSetup() {
    this._media = useMediaContext();
    this.$state.video.set(this._video);
  }
  onConnect() {
    const { crossorigin, canLoad } = this._media.$state, { src } = this.$state;
    if (this._video.parentNode !== this) {
      this.prepend(this._video);
    }
    effect(() => {
      setAttribute(this._video, "src", src());
      setAttribute(this._video, "crossorigin", crossorigin());
      setAttribute(this._video, "preload", canLoad() ? "auto" : "none");
    });
  }
  _createVideo() {
    return cloneTemplateContent(videoTemplate);
  }
}

class MediaTimeSliderElement extends Host(HTMLElement, TimeSlider) {
  static {
    this.tagName = "media-time-slider";
  }
}

class MediaSliderPreviewElement extends Host(HTMLElement, SliderPreview) {
  static {
    this.tagName = "media-slider-preview";
  }
}

class MediaVolumeSliderElement extends Host(HTMLElement, VolumeSlider) {
  static {
    this.tagName = "media-volume-slider";
  }
}

class MediaSliderChaptersElement extends Host(HTMLElement, SliderChapters) {
  constructor() {
    super(...arguments);
    this._template = null;
  }
  static {
    this.tagName = "media-slider-chapters";
  }
  onConnect() {
    requestScopedAnimationFrame(() => {
      if (!this.connectScope)
        return;
      const template = this.querySelector("template");
      if (template) {
        this._template = template;
        effect(this._renderTemplate.bind(this));
      }
    });
  }
  _renderTemplate() {
    if (!this._template)
      return;
    const elements = cloneTemplate(this._template, this.cues.length || 1);
    this.setRefs(elements);
  }
}

export { MediaChaptersRadioGroupElement as A, MediaSpeedRadioGroupElement as B, MediaQualityRadioGroupElement as C, MediaRadioElement as D, MediaRadioGroupElement as E, MediaSliderElement as F, MediaSliderThumbnailElement as G, MediaSliderValueElement as H, MediaSliderVideoElement as I, MediaTimeSliderElement as J, MediaSliderPreviewElement as K, MediaVolumeSliderElement as L, MediaCaptionsElement as M, MediaSliderChaptersElement as N, MediaGestureElement as a, MediaPosterElement as b, MediaThumbnailElement as c, MediaTimeElement as d, MediaControlsElement as e, MediaControlsGroupElement as f, MediaChapterTitleElement as g, MediaLayoutElement as h, MediaPlayButtonElement as i, MediaMuteButtonElement as j, MediaCaptionButtonElement as k, MediaFullscreenButtonElement as l, MediaPIPButtonElement as m, MediaSeekButtonElement as n, MediaToggleButtonElement as o, MediaLiveButtonElement as p, MediaTooltipElement as q, MediaTooltipTriggerElement as r, MediaTooltipContentElement as s, MediaAudioRadioGroupElement as t, MediaCaptionsRadioGroupElement as u, MediaMenuElement as v, MediaMenuButtonElement as w, MediaMenuPortalElement as x, MediaMenuItemElement as y, MediaMenuItemsElement as z };
