// vuepress-theme-hope/lib/client/components/AutoLink.js
import { defineComponent, h } from "vue";
import { AutoLink } from "vuepress/client";
import HopeIcon from "@theme-hope/components/HopeIcon";
export default defineComponent({
    name: "AutoLink",
    props: {
        /**
         * @description Autolink config
         */
        config: {
            type: Object,
            required: true,
        },
    },
    emits: ["focusout"],
    slots: Object,
    setup(props, { emit, slots }) {
        return () => {
            const { icon } = props.config;
            return h(AutoLink, {
                ...props, // Class needs to be merged manually
                onFocusout: () => emit("focusout"),
            }, {
                default: slots.default
                    ? () => slots.default()
                    : null,
                before: slots.before
                    ? () => slots.before()
                    : icon
                        ? () => h(HopeIcon, { icon })
                        : null,
                after: slots.after
                    ? () => slots.after()
                    : null,
            });
        };
    },
});
