// vuepress-theme-hope/lib/client/modules/sidebar/components/SidebarChild.js
import { isString } from "@vuepress/helper/client";
import { defineComponent, h } from "vue";
import { useRoute } from "vuepress/client";
import AutoLink from "@theme-hope/components/AutoLink";
import HopeIcon from "@theme-hope/components/HopeIcon";
import { isActiveSidebarItem } from "@theme-hope/modules/sidebar/utils/index";
import "../styles/sidebar-child.scss";
export default defineComponent({
    name: "SidebarChild",
    props: {
        /**
         * Sidebar item config
         *
         * 侧边栏项目配置
         */
        config: {
            type: Object,
            required: true,
        },
    },
    setup(props) {
        const route = useRoute();
        return () => isString(props.config.link)
            ? // If the item has link, render it as `<AutoLink>`
                h(AutoLink, {
                    class: [
                        "vp-sidebar-link",
                        `vp-sidebar-page`,
                        { active: isActiveSidebarItem(route, props.config, true) },
                    ],
                    exact: true,
                    config: props.config,
                })
            : // If the item only has text, render it as `<p>`
                h("p", props, [
                    h(HopeIcon, { icon: props.config.icon }),
                    props.config.text,
                ]);
    },
});
