// vuepress-theme-hope/lib/client/modules/sidebar/composables/setup.js
import { computedWithControl } from "@vueuse/core";
import { computed, inject, provide } from "vue";
import { usePageData, usePageFrontmatter, useRouteLocale, } from "vuepress/client";
import { useThemeLocaleData } from "@theme-hope/composables/index";
import { resolveSidebarItems } from "./resolveConfig.js";
export const sidebarItemsSymbol = Symbol(__VUEPRESS_DEV__ ? "sidebarItems" : "");
/**
 * Create sidebar items ref and provide as global computed in setup
 */
export const setupSidebarItems = () => {
    const frontmatter = usePageFrontmatter();
    const themeLocale = useThemeLocaleData();
    const page = usePageData();
    const routeLocale = useRouteLocale();
    // Get sidebar config from frontmatter > themeConfig
    const sidebarConfig = computed(() => frontmatter.value.home
        ? false
        : frontmatter.value.sidebar ?? themeLocale.value.sidebar ?? "structure");
    const headerDepth = computed(() => frontmatter.value.headerDepth ?? themeLocale.value.headerDepth ?? 2);
    const sidebarItems = computedWithControl(() => [
        sidebarConfig.value,
        headerDepth.value,
        page.value.path,
        __VUEPRESS_DEV__ ? page.value.headers : null,
    ], () => resolveSidebarItems({
        config: sidebarConfig.value,
        routeLocale: routeLocale.value,
        page: page.value,
        headerDepth: headerDepth.value,
    }));
    provide(sidebarItemsSymbol, sidebarItems);
};
/**
 * Inject sidebar items global computed
 */
export const useSidebarItems = () => {
    const sidebarItems = inject(sidebarItemsSymbol);
    if (!sidebarItems)
        throw new Error("useSidebarItems() is called without provider.");
    return sidebarItems;
};
